src/index.ts

```diff
@@ -138,20 +140,61 @@ class DebuggerHandler<
     widget: W,
     client: IClientSession | Session.ISession
   ): Promise<void> {
-    const debuggingEnabled = await debug.isAvailable(client);
-    if (!debug.model || !debuggingEnabled) {
-      return;
+    const updateHandler = () => {
+      void this._update(debug, widget, client);
+    };
+
+    // setup handler when the kernel changes
+    const kernelChangedHandler = this._kernelChangedHandlers[client.path];
+    if (kernelChangedHandler) {
+      client.kernelChanged.disconnect(kernelChangedHandler);
     }
+    client.kernelChanged.connect(updateHandler);
+    this._kernelChangedHandlers[client.path] = updateHandler;
+
+    // setup handler when the status of the kernel changes (restart)
+    // TODO: is there a better way to handle restarts?
+    let restarted = false;
+    const statusChanged = async () => {
+      // wait for the first `idle` status after a restart
+      if (restarted && client.status === 'idle') {
+        restarted = false;
+        return updateHandler();
+      }
+      // handle `starting`, `restarting` and `autorestarting`
+      if (client.status.endsWith('starting')) {
+        restarted = true;
+      }
+    };
```

Comment:
Not super satisfied with this logic.

It seems to be a bit tricky to really detect a kernel restart.

Reply:
An alternative would be to schedule the call to `updateHandler` with a small time delay.